core-api/src/telemetry/metrics.rs

```diff
@@ -197,14 +164,92 @@ pub trait Gauge: Send + Sync {
     fn record(&self, value: u64, attributes: &MetricAttributes);
 }
 
+#[derive(Debug, Clone)]
+pub enum MetricEvent<I> {
+    Create {
+        params: MetricParameters,
+        /// One you receive this event, call `set` on this with the initialized instrument reference
+        populate_into: LazyBufferInstrument<I>,
+        kind: MetricKind,
+    },
+    CreateAttributes {
+        /// One you receive this event, call `set` on this with the initialized attributes
+        populate_into: BufferAttributes,
+        /// If not `None`, use these already-initialized attributes as the base (extended with
+        /// `attributes`) for the ones you are about to initialize.
+        append_from: Option<BufferAttributes>,
+        attributes: Vec<MetricKeyValue>,
+    },
+    Update {
+        instrument: LazyBufferInstrument<I>,
+        attributes: BufferAttributes,
+        update: MetricUpdateVal,
+    },
+}
+#[derive(Debug, Clone, Copy)]
+pub enum MetricKind {
+    Counter,
+    Gauge,
+    Histogram,
+}
+#[derive(Debug, Clone, Copy)]
+pub enum MetricUpdateVal {
+    // Currently all deltas are natural numbers
+    Delta(u64),
+    // Currently all values are natural numbers
+    Value(u64),
+}
+
+pub trait MetricCallBufferer<I>: Send + Sync {
+    fn retrieve(&self) -> Vec<MetricEvent<I>>;
+}
+
+/// A lazy reference to some metrics buffer attributes
+pub type BufferAttributes = LazyRef<Arc<dyn CustomMetricAttributes + 'static>>;
+
+/// Types lang uses to contain references to its lang-side defined instrument references must
+/// implement this marker trait
+pub trait BufferInstrumentRef {}
+/// A lazy reference to a metrics buffer instrument
+pub type LazyBufferInstrument<T> = LazyRef<Arc<T>>;
```

Comment:
Can we put a type constraint on `T` that it has to implement `BufferInstrumentRef`? Or maybe `I` can have that constraint above? Or rather, is it at least expected that `T` implements `BufferInstrumentRef`?

Reply:
EDIT: Nevermind, I see this constraint is now available on the `CoreMeter` impl